test/server.go

```diff
@@ -0,0 +1,253 @@
+type AttributesServer struct {
+	// GlobalDict controls the known global dictionary for attribute processing.
+	GlobalDict map[string]int32
+
+	// GenerateGRPCError instructs the server whether or not to fail-fast with
+	// an error that will manifest as a GRPC error.
+	GenerateGRPCError bool
+
+	// CheckResponseStatus controls what rpc.Status message is returned in the
+	// Preconditions section of CheckResponse.
+	CheckResponseStatus rpc.Status
+
+	// ReferencedAttributes controls what set of attributes are returned by
+	// the server for the request (for cache testing).
+	ReferencedAttributes mixerpb.ReferencedAttributes
+
+	// QuotaResults enables overriding the default quota results behavior of
+	// the server.
+	QuotaResults map[string]mixerpb.CheckResponse_QuotaResult
+
+	// CheckAttributes is the channel on which the attribute bag generated by
+	// the server for Check() requests is sent.
+	CheckAttributes chan attribute.Bag
+
+	// QuotaDispatches is the channel on which the quota information generated by
+	// the server for Check() requests is sent.
+	QuotaDispatches chan QuotaDispatchInfo
+
+	// ReportAttributes is the channel on which the attribute bag generated by
+	// the server for Report() requests is sent.
+	ReportAttributes chan attribute.Bag
+}
+
+// NewAttributesServer creates an AttributesServer. All channels are set to
+// default length.
+func NewAttributesServer() *AttributesServer {
+	list := attributes.GlobalList()
+	globalDict := make(map[string]int32, len(list))
+	for i := 0; i < len(list); i++ {
+		globalDict[list[i]] = int32(i)
+	}
+
+	return &AttributesServer{
+		globalDict,
+		false,
+		rpc.Status{Code: int32(rpc.OK)},
+		mixerpb.ReferencedAttributes{},
+		make(map[string]mixerpb.CheckResponse_QuotaResult),
+		make(chan attribute.Bag),
+		make(chan QuotaDispatchInfo),
+		make(chan attribute.Bag),
+	}
+
+}
+
+// Check sends a copy of the protocol buffers attributes wrapper for the preconditions
+// check as well as for each quotas check to the CheckAttributes channel. It also
+// builds a CheckResponse based on server fields. All channel sends timeout to
+// prevent problematic tests from blocking indefinitely.
+func (a *AttributesServer) Check(ctx context.Context, req *mixerpb.CheckRequest) (*mixerpb.CheckResponse, error) {
+
+	if a.GenerateGRPCError {
+		return nil, errors.New("error handling check call")
+	}
+
+	requestBag := attribute.NewProtoBag(&req.Attributes, a.GlobalDict, attributes.GlobalList())
+	defer requestBag.Done()
+
+	// avoid blocked go-routines in testing
+	select {
+	case a.CheckAttributes <- attribute.CopyBag(requestBag): // no-op
+	case <-time.After(3 * time.Second):
+		return nil, errors.New("timeout trying to send attribute bag")
+	}
+
+	status := a.CheckResponseStatus
+
+	resp := &mixerpb.CheckResponse{
+		Precondition: mixerpb.CheckResponse_PreconditionResult{
+			Status:               status,
+			ValidUseCount:        DefaultValidUseCount,
+			Attributes:           req.Attributes,
```

Comment:
My testing code may want to control returning attributes. we should have a way for testing code to set it.